### PR TITLE
feat(actions): expose pending comment before persistence

### PR DESCRIPTION
**Pull request by tomcasaburi**
## Summary

- expose a provisional account comment before local persistence completes
- guard the callback after abandon or unmount and isolate callback failures
- document and test the new public callback

## Verification

- `yarn build`
- `yarn lint`
- `yarn type-check`
- `yarn knip`
- `yarn test` (1,167 passed, 6 skipped)
- `yarn test:coverage` (1,167 passed, 6 skipped)

---

> [!NOTE]
> **Medium Risk**
> Touches publish/abandon lifecycle and account comment indexing with many edge-case tests; behavior change for clients that relied only on hook `index` after persistence, but failures in the new callback are isolated from publish success.
> 
> **Overview**
> Adds optional **`onPendingComment`** on `usePublishComment` / `publishComment` so apps can navigate or render a provisional account comment **before** IndexedDB persistence finishes, with a **second call** if a concurrent local deletion shifts the index.
> 
> **Store layer:** fires the callback immediately with a cloned snapshot (mutations in the callback don’t affect persisted data); publish continues if the callback throws or rejects; strips `onPendingComment` before `pkc.createComment`.
> 
> **Hook layer:** per-publish request guards so `onPendingComment` and hook state updates stop after **abandon**; **deferred abandon** waits for a pending index then deletes; challenge callbacks still run after unmount while publish guards block stale UI updates; `abandonPublish` won’t delete a prior comment when abandoning a second publish.
> 
> Docs (`README`, `llms-full.txt`) and changelog snippets updated; validator accepts optional `onPendingComment`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bec3d4789ee505a54bdce947ffc372f0ea7e72f5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added an optional callback for handling provisional comments immediately, before local persistence completes.
  * Provisional comment details and their index can now be used for immediate navigation or rendering.

* **Bug Fixes**
  * Publishing continues safely when the callback fails.
  * Callbacks are no longer triggered after publishing is abandoned or the related component is removed.
  * Added validation for invalid callback values.

* **Documentation**
  * Updated publishing examples and generated reference documentation with callback behavior and recent changelog entries.

### CODE REVIEW

File: src/stores/accounts/accounts-actions.ts [hunk 1]
```diff
@@ -1142,8 +1151,26 @@ export const publishComment = async (
...
+      publishCommentOptions.onPendingComment?.(accountCommentIndex, pendingCommentSnapshot),
+    ).catch(reportOnPendingCommentError);
+  } catch (error) {
+    reportOnPendingCommentError(error);
+  }
+  const pendingCommentIndex = await saveCreatedAccountComment(createdAccountComment);
+  if (pendingCommentIndex === undefined) {
+    return createdAccountComment;
+  }
+  publishCommentOptions._onPendingCommentIndex?.(pendingCommentIndex, createdAccountComment);
```
[the remark this answers is not in the document]
Comment (tomcasaburi) [reply src/stores/accounts/accounts-actions.ts:1180]: Fixed in fda87877. After persistence determines the shifted slot, the returned account comment now carries that persisted index. The shifted-index regression test asserts the internal callback, stored comment, and resolved return value all use the same index.

File: src/hooks/actions/actions.ts [hunk 2]
```diff
@@ -212,6 +212,12 @@ export function usePublishComment(options?: UsePublishCommentOptions): UsePublis
   const publishRequestIdRef = useRef(0);
   const activePublishRequestIdRef = useRef<number | undefined>(undefined);
   const guardActive = () => activePublishRequestIdRef.current !== undefined;
+  useEffect(
+    () => () => {
+      activePublishRequestIdRef.current = undefined;
```
[the remark this answers is not in the document]
Comment (tomcasaburi) [reply src/hooks/actions/actions.ts]: Fixed in ce4613dc. Publish callbacks now continue after the hook unmounts, while React state setters are gated by the mounted ref. The regression test covers a challenge arriving after unmount and abandonment deleting the pending comment.

File: src/stores/accounts/accounts-actions.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomcasaburi) [reply src/stores/accounts/accounts-actions.ts:1180]: The public callback intentionally reports the current in-memory slot before persistence so consumers can navigate immediately. An unrelated concurrent deletion can shift array indexes after that callback has already been delivered; representing that later remap would require a stable pending-publication identifier and a wider API/route change. The post-persistence internal callback and final return now both carry the shifted slot, and 5chan serializes abandonment before starting the next publish. Treating the cross-tab remap case as outside this callback's current contract.

File: src/hooks/actions/actions.ts [hunk 3]
```diff
@@ -211,83 +216,145 @@ export function usePublishComment(options?: UsePublishCommentOptions): UsePublis
...
+    }
+    if (requestId !== undefined) {
+      let resolve!: () => void;
+      let reject!: (error: unknown) => void;
+      const promise = new Promise<void>((resolvePromise, rejectPromise) => {
+        resolve = resolvePromise;
+        reject = rejectPromise;
+      });
+      deferredAbandonsRef.current.set(requestId, { promise, resolve, reject });
+      await promise;
```
[the remark this answers is not in the document]
Comment (tomcasaburi) [reply src/hooks/actions/actions.ts:362]: Fixed in 5c28dd66. Terminal publish failures now clear the matching active request in a `finally` block, including when error reporting throws. The failure regression test now calls `abandonPublish()` afterward and verifies it resolves instead of installing an unreachable deferred cancellation.

File: src/stores/accounts/accounts-actions.ts [hunk 4]
```diff
@@ -1142,8 +1151,27 @@ export const publishComment = async (
...
+    log.error("accountsActions.publishComment onPendingComment callback error", {
+      accountId: account.id,
+      accountCommentIndex,
+      error,
+    });
+  };
+  try {
+    const pendingCommentSnapshot = utils.clone(createdAccountComment) as AccountComment;
+    void Promise.resolve(
+      publishCommentOptions.onPendingComment?.(accountCommentIndex, pendingCommentSnapshot),
```
[the remark this answers is not in the document]
Comment (tomcasaburi) [reply src/stores/accounts/accounts-actions.ts]: Fixed in bec3d478. `onPendingComment` still fires immediately, and if persistence shifts the pending slot it fires again with a cloned corrected-index snapshot. The shifted-index race test now asserts callbacks at indexes 1 then 0, and the README/type docs describe the correction contract.
